syllabus/curricula/selfplay.py

```diff
@@ -0,0 +1,155 @@
+            spaces.Discrete(1)
+        )  # SelfPlay can only return agent_id = 0
+
+    def update_agent(self, agent: Agent) -> Agent:
+        self.agent = deepcopy(agent).to(self.device)
+
+    def get_opponent(self, agent_id) -> Agent:
+        if agent_id is None:
+            agent_id = 0
+        assert (
+            agent_id == 0
+        ), f"Self play only tracks the current agent. Expected agent id 0, got {agent_id}"
+        return self.agent
+
+    def sample(self, k=1):
+        return 0
+
+
+class FictitiousSelfPlay(Curriculum):
+    def __init__(
+        self,
+        agent: Agent,
+        device: str,
+        storage_path: str,
+        max_agents: int,
+    ):
+        self.name = "FSP"
+        self.device = device
+        self.storage_path = storage_path
+        if not os.path.exists(self.storage_path):
+            os.makedirs(self.storage_path, exist_ok=True)
+
+        self.n_stored_agents = 0
+        self.max_agents = max_agents
+        self.task_space = TaskSpace(spaces.Discrete(self.max_agents))
+        self.update_agent(agent)  # creates the initial opponent
+
+    def update_agent(self, agent):
+        """Saves the current agent instance to a pickle file."""
+        if self.n_stored_agents < self.max_agents:
+            # TODO: define the expected behaviour when the limit is exceeded
+            joblib.dump(
+                agent,
+                filename=f"{self.storage_path}/{self.name}_agent_checkpoint_{self.n_stored_agents}.pkl",
+            )
+            self.n_stored_agents += 1
+
+    def get_opponent(self, agent_id) -> Agent:
+        """Loads an agent from the buffer of saved agents."""
+        return joblib.load(
+            f"{self.storage_path}/{self.name}_agent_checkpoint_{agent_id}.pkl"
+        ).to(self.device)
+
+    def sample(self, k=1):
+        return np.random.randint(self.n_stored_agents)
+
+
+class PrioritizedFictitiousSelfPlay(Curriculum):
+    def __init__(
+        self,
+        agent: Agent,
+        device: str,
+        storage_path: str,
+        max_agents: int,
+    ):
+        self.name = "PFSP"
+        self.device = device
+        self.storage_path = storage_path
+        if not os.path.exists(self.storage_path):
+            os.makedirs(self.storage_path, exist_ok=True)
+
+        self.n_stored_agents = 0
+        self.max_agents = max_agents
+        self.task_space = TaskSpace(spaces.Discrete(self.max_agents))
+        self.update_agent(agent)  # creates the initial opponent
+        self.history = {
+            i: {
+                "winrate": 0,
+                "n_games": 0,
+            }
+            for i in range(self.max_agents)
+        }
+
+    def update_agent(self, agent) -> None:
+        """
+        Saves the current agent instance to a pickle file and update
+        its priority.
+        """
+        if self.n_stored_agents < self.max_agents:
+            # TODO: define the expected behaviour when the limit is exceeded
```

Comment:
We probably should delete or overwrite agents when this happens.

Reply:
Then I suggest this: 
```python
joblib.dump(
            agent,
            filename=f"{self.storage_path}/{self.name}_agent_checkpoint_{self.current_agent_index % self.max_agents}.pkl",
        )
self.current_agent_index += 1
```

Reply:
If we go this route, you should save the most recent agent to a file some.where. Since it won't be obvious from the filenames. It might be better to just delete a file and write a new one, but this is fine for now.